src/transaction/atomic_replication_helper.cpp

```diff
@@ -268,35 +269,35 @@ namespace cublog
   atomic_replication_helper::atomic_log_sequence::add_atomic_replication_log (THREAD_ENTRY *thread_p,
       log_lsa record_lsa, LOG_RCVINDEX rcvindex, VPID vpid)
   {
-    m_log_vec.emplace_back (record_lsa, vpid, rcvindex);
-    const auto page_map_it = m_page_map.find (vpid);
-    if (page_map_it == m_page_map.cend ())
+    PAGE_PTR page_p = nullptr;
+    // bookkeeping fixes page, keeps all info regarding how the page was fixed (either
+    // regular fix or ordered fix) and only returns back the pointer to the page
+    const int err_code = m_page_ptr_bookkeeping.fix_page (thread_p, vpid, rcvindex, page_p);
+    if (err_code != NO_ERROR)
       {
-	int error_code = m_log_vec.back ().fix_page (thread_p);
-	if (error_code != NO_ERROR)
-	  {
-	    // failing to fix the page just leaves it unfixed and does not affect overall
-	    // functioning of the atomic replication sequence;
-	    // therefore, remove the log and move on
-	    m_log_vec.pop_back ();
-
-	    // TODO:
-	    //  - what happens if there is more than one log record pertaining to the same page
-	    //    in an atomic sequnce and, for example, for the first log record, the page fails
-	    //    to be fixed but succeeds for the second log record
-	    //  - what if, while the atomic sequence is in progress with a page having failed to
-	    //    be fixed, a client transactions manages to fix the page; IOW, how is the progress
-	    //    of the "highest processed LSA" working wrt atomic replication sequences
-
-	    return error_code;
-	  }
-	m_page_map.emplace (vpid,  m_log_vec.back ().get_page_ptr ());
+	// failing to fix the page just leaves it unfixed and does not affect overall
+	// functioning of the atomic replication sequence;
+	er_log_debug (ARG_FILE_LINE, "[ATOMIC_REPL] heap page cannot be fixed, cannot add new log record"
+		      "with LSA %lld|%d to atomic sequences started at LSA %lld|%d\n",
+		      LSA_AS_ARGS (&record_lsa), LSA_AS_ARGS (&m_start_lsa));
+
+	// TODO:
+	//  - what happens if there is more than one log record pertaining to the same page
```

Comment:
Applying only a subset of log records of an atomic change, does leave the DB in an inconsistent state.
Can we manage this case by force evacuating the page from page buffer (and then the replication or user transactions will trigger a  fetch from PS) ?

Reply:
In the first phase I will implement and extension to the bookkeeping to:
- keep track if a page failed to be fixed
- and a subsequent request to the same page - within the same atomic sequence - succeeds

At that point we'll at least throw an assert and, if that ever happens, re-asses at that point.

From what I saw, multiple modifications to the same page do occur when there are nested atomic replications.
But once we implement the simplification of how we handle atomic sequences - in the sense that we'll apply everything before the postpone as a standalone atomic sequence - this should not be a common case anymore.